### ISSUE TITLE
Regression tests falsely marked as failed when prior failed attempt finishes after next successful attempt starts

### ISSUE THREAD

**Issue by SamuelTrahanNOAA** [ufs-community/ufs-weather-model#1309]
## Description

Sometimes the regression test system will falsely mark a test as failed if it was resubmitted. This is due to a race condition described below. The easiest solution is for the jobs to delete their `fail_(number)` files at the end of the job, not just the beginning.

### How the System is Designed

The regression test system makes `fail_*` files to indicate a regression test failed. The intended order of operations in the job is this:

1. First job starts.
2. First job deletes the `fail_(number)` file
3. First job runs the test, and the test fails.
4. First job makes the `fail_(number)` file to indicate the test failed.
5. First job exits.
6. Job is resubmitted.
7. Second job deletes the `fail_(number)` file
8. Second job runs the test
9. Second job's test succeeded, and the job exits

### What can go Wrong

The trouble is, this can happen in a different order if the first job is killed or preempted. In that situation, a SIGINT handler makes the fail file, and that can happen after the job completed. 

1. First job starts.
2. First job deletes the `fail_(number)` file
3. First job runs the test, but the test is interrupted abnormally
4. First job is killed
5. Job is resubmitted
6. Second job starts.
7. Second job deletes the `fail_(number)` file
8. First job's SIGINT handler creates the `fail_(number)` file
9. First job exits.
10. Second job runs the test
11. The test succeeded, and the job exits
12. The `fail_(number)` file is still present because the second job did not delete it.

### The Solution

The simplest solution is for the job to delete the `fail_(number)` file at the end if the test succeeded. Then the order of operations is:

1. First job starts.
2. First job deletes the `fail_(number)` file
3. First job runs the test, but the test is interrupted abnormally
4. First job is killed
5. Job is resubmitted
6. Second job starts.
7. Second job deletes the `fail_(number)` file
8. First job's SIGINT handler creates the `fail_(number)` file
9. First job exits.
10. Second job runs the test
11. Second job's test succeeded
12. Second job deletes the `fail_(number)` file.
13. The regression test system considers the test successful because the `fail_(number)` file is absent.

### To Reproduce:
This is frequently a problem on Jet, but other systems can have it too.
1. Run the regression test system many times.
2. Witness the problem.

**Comment by SamuelTrahanNOAA** [comment]
Seeking comment from @DusanJovic-NOAA, @junwang-noaa, @BrianCurtis-NOAA, and @jkbk2004 on this bug before I fix it.

**Comment by DusanJovic-NOAA** [comment]
Suggested solution look reasonable to me. I wonder if there is a simple way to avoid race condition you described.

**Comment by SamuelTrahanNOAA** [comment]
The fix seems to be copying this code in `run_test.sh`:

```
cd ${PATHRT}
OPNREQ_TEST=${OPNREQ_TEST:-false}
if [[ ${OPNREQ_TEST} == true ]]; then
  rm -f fail_opnreq_test_${TEST_NR}
else
  rm -f fail_test_${TEST_NR}
fi
```

To just before the timestamp:

```
echo " $( date +%s ), ${NODES}" >> ${LOG_DIR}/job_${JOB_NR}_timestamp.txt
```

Inside an if block:

```
if [ "$test_status" != FAIL ] ; then
  cd ${PATHRT}
  OPNREQ_TEST=${OPNREQ_TEST:-false}
  if [[ ${OPNREQ_TEST} == true ]]; then
    rm -f fail_opnreq_test_${TEST_NR}
  else
    rm -f fail_test_${TEST_NR}
  fi
fi

echo " $( date +%s ), ${NODES}" >> ${LOG_DIR}/job_${JOB_NR}_timestamp.txt
```

I could also put it in a function, and call the function from both places. That would be cleaner but equivalent.

**Comment by SamuelTrahanNOAA** [comment]
> Suggested solution look reasonable to me. I wonder if there is a simple way to avoid race condition you described.

That would be highly system-dependent since the job cleanup depends on the chosen batch system and is usually configurable in any batch system. 

I do see two other options, but neither are clean nor reliable:

1. Use a lock directory to guard the flag files, but that risks freezing the second job if something prevents the first job from deleting the lock directory.

2. Trust the metascheduler (ecFlow, Rocoto, shell, etc.) to correctly report the success of failure status of a job. In my experience, this is not a reliable option. Many things can go wrong, as there are many corner cases and ways the metaschedulers can fail.

**Comment by DusanJovic-NOAA** [comment]
Ok. Then I suggest to fix this issue as you suggested. Always delete any existing fail_*** file at the beginning of the run script and at the end if test is successful.

**Comment by BrianCurtis-NOAA** [comment]
Thought: @DusanJovic-NOAA would it be simpler to add a random number, say: `RANNUM=$RANDOM` and use that in the fail_test file name?

**Comment by SamuelTrahanNOAA** [comment]
Using a random number won't help because you'll still have the old failure files sitting around. Actually, that'll give you a false failure every time a job is resubmitted.

**Comment by BrianCurtis-NOAA** [comment]
> Using a random number won't help because you'll still have the old failure files sitting around. Actually, that'll give you a false failure every time a job is resubmitted.

Yes, true. rt.sh uses fail_test so we don't want to change multiple files if we can do it in one. I agree this approach is sufficient.

**Comment by SamuelTrahanNOAA** [comment]
My proposed bug fix is in #1312.
